python/lsst/meas/astrom/utils.cc

```diff
@@ -53,282 +54,295 @@ read_column(fitstable_t* tag,
-             stargal = static_cast<bool*>(malloc(nstars));
-            if (!stargal) {
-                throw LSST_EXCEPT(lsst::pex::exceptions::NotFoundError,
-                                  str(boost::format("Unable to read data for %s from %s") %
-                                      stargalcol % ind->indexname));
+                 msg += " ].  You may need to edit the $ASTROMETRY_NET_DATA_DIR/andConfig.py file to set idColumn=None, etc.";
+                 throw LSST_EXCEPT(lsst::pex::exceptions::NotFoundError, msg);
             }
-            for (int j=0; j<nstars; j++) {
-                stargal[j] = (sg[j] > 0);
+
+            if (idCol) {
+                id = static_cast<int64_t*>(fitstable_read_column_inds(tag, idCol, i64, starinds, nstars));
+                if (!id) {
+                    throw LSST_EXCEPT(lsst::pex::exceptions::NotFoundError,
+                        str(boost::format("Unable to read data for %s from %s") % idCol % ind->indexname));
+                }
             }
-            free(sg);
-         }
-         if (varcol) {
-             var = static_cast<bool*>(fitstable_read_column_inds(tag, varcol, boo, starinds, nstars));
-             if (!var) {
-                throw LSST_EXCEPT(lsst::pex::exceptions::NotFoundError,
-                                  str(boost::format("Unable to read data for %s from %s") %
-                                      varcol % ind->indexname));
+            if (id && uniqueIds) {
+                // remove duplicate IDs.
+
+                // FIXME -- this shouldn't be necessary once we get astrometry_net 0.40
+                // multi-index functionality in place.
+
+                if (uids.empty()) {
+                    uids = std::set<boost::int64_t>(id, id+nstars);
+                } else {
+                    int nkeep = 0;
+                    for (int i=0; i<nstars; i++) {
+                        //std::pair<std::set<boost::int64_t>::iterator, bool> 
+                        if (uids.insert(id[i]).second) {
+                            // inserted; keep this one.
+                            if (nkeep != i) {
+                                // compact the arrays.
+                                starinds[nkeep] = starinds[i];
+                                radecs[nkeep*2+0] = radecs[i*2+0];
+                                radecs[nkeep*2+1] = radecs[i*2+1];
+                                id[nkeep] = id[i];
+                            }
+                            nkeep++;
+                        } else {
+                            // did not insert (this id has already been found);
+                            // drop this star.
+                        }
+                    }
+                    nstars = nkeep;
+                    // if they were all duplicate IDs...
+                    if (nstars == 0) {
+                        free(starinds);
+                        free(radecs);
+                        free(id);
```

Comment:
I know this isn't your code, but we *really* need to get rid of this C-style memory management, and replace it with smart pointers, especially considering that we have some exceptions being thrown in this code, and that could lead to memory leaks here.  Could you open an issue to deal with this in the future?